runtime/memory.c

```diff
@@ -557,16 +563,20 @@ CAMLexport caml_stat_block caml_stat_resize_noexc(caml_stat_block b, asize_t sz)
     return realloc(b, sz);
   else {
     struct pool_block *pb = get_pool_block(b);
-    struct pool_block *pb_new = realloc(pb, sz + SIZEOF_POOL_BLOCK);
-    if (pb_new == NULL) return NULL;
-
-    /* Relinking the new block into the ring in place of the old one */
-    caml_plat_lock(&pool_mutex);
```

Comment:
Naive question: wouldn't it suffice, to fix the race, to widen the critical section by acquiring the pool mutex right before the `realloc` call? I guess that `realloc` could block / take some time so we don't want to hold the lock at this point?

Reply:
Yes, realloc can take a while (it may need to copy the old block in the new block), and I prefer it if the critical section covers only a few instructions, so that contention is very unlikely.